ignite/handlers/early_stopping.py

```diff
@@ -5,47 +5,57 @@
 
-__all__ = ["EarlyStopping"]
+__all__ = ["NoImprovementHandler", "EarlyStopping"]
 
 
-class EarlyStopping(Serializable):
-    """EarlyStopping handler can be used to stop the training if no improvement after a given number of events.
-
+class NoImprovementHandler(Serializable):
+    """NoImprovementHandler is a generalised version of Early stopping where you can define what should
+    happen if no improvement occurs after a given number of events.
     Args:
-        patience: Number of events to wait if no improvement and then stop the training.
+        patience: Number of events to wait if no improvement and then call stop_function.
         score_function: It should be a function taking a single argument, an :class:`~ignite.engine.engine.Engine`
             object, and return a score `float`. An improvement is considered if the score is higher.
+        pass_function: It should be a function taking a single argument, the trainer
+            object, and defines what to do in the case when the stopping condition is not met.
+        stop_function: It should be a function taking a single argument, the trainer
+            object, and defines what to do in the case when the stopping condition is met.
         trainer: Trainer engine to stop the run if no improvement.
         min_delta: A minimum increase in the score to qualify as an improvement,
             i.e. an increase of less than or equal to `min_delta`, will count as no improvement.
         cumulative_delta: It True, `min_delta` defines an increase since the last `patience` reset, otherwise,
             it defines an increase after the last event. Default value is False.
-
     Examples:
         .. code-block:: python
 
-            from ignite.engine import Engine, Events
-            from ignite.handlers import EarlyStopping
+            #Example where if the score doesn't improve a user defined value `alpha` is doubled.
 
+            from ignite.engine import Engine, Events
+            from ignite.handlers import NoImprovementHandler
             def score_function(engine):
                 val_loss = engine.state.metrics['nll']
                 return -val_loss
+            def pass_function(engine):
+                pass
+            def stop_function(trainer):
+                trainer.state.alpha *= 2
+
+            trainer = Engine(do_nothing_update_fn)
+            trainer.state_dict_user_keys.append("alpha")
+            trainer.state.alpha = 0.1
+
+            h = NoImprovementHandler(patience=3, score_function=score_function, pass_function=pass_function,
+                                     stop_function=stop_function, trainer=trainer)
 
-            handler = EarlyStopping(patience=10, score_function=score_function, trainer=trainer)
             # Note: the handler is attached to an *Evaluator* (runs one epoch on validation dataset).
             evaluator.add_event_handler(Events.COMPLETED, handler)
-
     """
 
-    _state_dict_all_req_keys = (
```

Comment:
If we want to save NoImprovementHandler's internal state we have to keep `_state_dict_all_req_keys`, IMO